### PR TITLE
feat: migrate to v3, allow for embedded files and different database types

### PR DESCRIPTION
**Pull request by woehrl01**
Migrated API for compatibility with Coraza v3.

**Changes:**
- Transitioned to embedded database due to removal of `RegisterDirectivePlugin`.
  - Enhances compatibility with [coraza-proxy-wasm](https://github.com/corazawaf/coraza-proxy-wasm/issues/234).
- Introduced capability to select database type (city or country).
- Upgraded to Go 1.20.

Awaiting your feedback and suggestions.

### CODE REVIEW

**Comment by jcchavezs** [comment]
Thanks @woehrl01 do you mind also adding a readme and example usage? Also we could add some tests.

**Comment by woehrl01** [comment]
@jcchavezs absolutely, I just added tests, github action and a readme.

**Review by M4tteoP — APPROVED** [review]
Many thanks for the contribution @woehrl01! It looks good to me, let's see if @jcchavezs and @jptosso want to take a second look

**Comment by woehrl01** [comment]
Thank you @M4tteoP . Any feedback from your side @jcchavezs @jptosso ? Thanks!

File: operators.go [hunk 1]
```diff
@@ -1,42 +1,53 @@
...
-	geoip "github.com/oschwald/geoip2-golang"
+	"github.com/corazawaf/coraza/v3/collection"
+	"github.com/corazawaf/coraza/v3/experimental/plugins/plugintypes"
+	"github.com/corazawaf/coraza/v3/types/variables"
+	"github.com/oschwald/geoip2-golang"
 )
 
-type geoLookup struct{}
+// this acts as a shim for the geoip2.Reader struct so we can mock it in tests
+type geoIPReaderShim interface {
```
Comment (jcchavezs) [inline operators.go]: Do we really need the Shim suffix? can we just use `geoIPReader`, same for others.

File: operators.go [hunk 2]
```diff
@@ -1,42 +1,53 @@
...
-		return false
+// this acts as a shim for the plugintypes.TransactionState struct so we can mock it in tests
+type txShim interface {
+	GetGeoCollection(tx txShim) (collectionShim, error)
+}
+
+type txShimmer struct {
+	tx plugintypes.TransactionState
+}
+
```
Comment (jcchavezs) [inline operators.go:34]: I believe we should add the compile time type assertion to make sure the implementation works.

```
var txShim = (*txShimmer)(nil)
```

File: operators_logic.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+	return true, nil
+}
+
+func (o *geo) executeEvaluationInternal(tx txShim, value string) (bool, error) {
+	col, err := tx.GetGeoCollection(tx)
+	if err != nil {
+		return false, err
+	}
+	ip := net.ParseIP(value)
```
Comment (jcchavezs) [inline operators_logic.go:69]: I wonder two things here for the sake of early return:
1. if the valye is empty, shall we continue?
2. if the IP is private, shall we continue?

File: plugin.go [hunk 4]
```diff
@@ -1,12 +1,24 @@
...
+	"github.com/oschwald/geoip2-golang"
 )
 
-func init() {
-	operators.RegisterPlugin("geoLookup", func() coraza.RuleOperator { return new(geoLookup) })
-	seclang.RegisterDirectivePlugin("secGeoLookupDb", directiveSecGeoLookupDb)
+func RegisterGeoDatabase(database []byte, databaseType string) {
+	db, err := geoip2.FromBytes(database)
+	if err != nil {
+		panic(err)
```
Comment (jcchavezs) [inline plugin.go]: not sure about this panic. I believe we don't want to turn down the platform.

File: operators_logic.go (hunk 3 above)
Comment (woehrl01) [reply operators_logic.go:69]: Good question, I would say for 1. We should not continue, for the 2. We should.

File: plugin.go (hunk 4 above)
Comment (woehrl01) [reply plugin.go]: Fair point, I return the error now, and leave the handling up to the caller.

**Comment by woehrl01** [comment]
Thank you @jcchavezs ! I made the suggested modifications, and also added a third entry point to allow passing a custom handler to the creation phase, this should allow users some additional flexibility to use that plugin.

File: operators_logic.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (o *geo) executeEvaluationInternal(tx Transaction, value string) (bool, error) {
+	col, err := tx.GetGeoCollection(tx)
+	if err != nil {
+		return false, err
+	}
+	ip := net.ParseIP(value)
+	if ip == nil {
+		return false, fmt.Errorf("invalid ip")
```
Comment (jcchavezs) [inline operators_logic.go]:
```suggestion
		return false, fmt.Errorf("invalid ip %q", value)
```

File: operators_logic.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	col.Set("country_continent", []string{r.Continent.Names["en"]})
+
+	return true, nil
+}
+
+func (o *geo) executeEvaluationInternal(tx Transaction, value string) (bool, error) {
+	col, err := tx.GetGeoCollection(tx)
+	if err != nil {
+		return false, err
+	}
```
Comment (jcchavezs) [inline operators_logic.go:68]:
```suggestion
	}
	
	if len(value) == 0 {
		return false, errors.New("empty IP")
	}
```

File: operators.go [hunk 7]
```diff
@@ -1,42 +1,55 @@
...
-	db, ok := tx.Waf.Config.Get("geoip", nil).(*geoip.Reader)
-	if !ok || db == nil {
-		return false
+// this acts as a shim for the collection.Map struct so we can mock it in tests
+type mapCollection interface {
+	Set(key string, values []string)
+}
+
+// this acts as a shim for the plugintypes.TransactionState struct so we can mock it in tests
+type Transaction interface {
```
Comment (jcchavezs) [inline operators.go]: Does this need to be public?

Comment (woehrl01) [reply operators.go]: No, I didn't had a better name 😅 which is lowercase which wouldn't conflict with imports or existing names, any suggestions?

Comment (woehrl01) [reply operators.go]: made it private and changed some names

File: operators_logic.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	return func(options plugintypes.OperatorOptions) (plugintypes.Operator, error) {
+		return newGeolookup(options, db, databaseType)
+	}
+}
+
+func newGeolookup(options plugintypes.OperatorOptions, db *geoip2.Reader, databaseType string) (plugintypes.Operator, error) {
+	return &geo{db: db, dbtype: databaseType}, nil
+}
+
+func (o *geo) ApplyVariablesCity(col mapCollection, ip net.IP) (bool, error) {
```
Comment (jcchavezs) [inline operators_logic.go]: I wonder if this should be public.

File: operators_logic.go (hunk 6 above)
Comment (jcchavezs) [reply operators_logic.go:68]: any particular reason to close this @woehrl01 ?

File: operators_logic.go (hunk 8 above)
Comment (woehrl01) [reply operators_logic.go]: you're right, I just went through all methods and made them private where appropriate. Thanks!

File: operators_logic.go (hunk 6 above)
Comment (woehrl01) [reply operators_logic.go:68]: no, looks like I somehow closed it together with your other suggestion, I just applied your suggestion in a dedicated commit.

File: operators_logic.go (hunk 3 above)
Comment (jcchavezs) [reply operators_logic.go:69]: CC @fzipi @jptosso

**Review by jcchavezs — APPROVED** [review]
LGTM I would just like to hear about private IPs

File: operators_logic.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/corazawaf/coraza/v3/experimental/plugins/plugintypes"
+	"github.com/oschwald/geoip2-golang"
+)
+
+type geo struct {
+	db     geoIPReader
+	dbtype string
+}
+
+func newGeolookupCreator(db *geoip2.Reader, databaseType string) func(options plugintypes.OperatorOptions) (plugintypes.Operator, error) {
```
Comment (jcchavezs) [inline operators_logic.go]: nit: newGeolookupBuilder

**Review by anuraaga — APPROVED** [review]

File: operators_logic.go (hunk 3 above)
Comment (fzipi) [reply operators_logic.go:69]: I think for both cases we should not continue. Also, there might be more border cases that we might want to take into account. What about [`0.0.0.0`](https://pkg.go.dev/net/netip#IPv4Unspecified) or `255.255.255.0`?

Comment (woehrl01) [reply operators_logic.go:69]: @fzipi I just added the skipping of specific ip ranges. I'm not sure if we should return `false` or `true` in that case. `false` means the filter failed, and the request gets denied, correct?

**Comment by jcchavezs** [comment]
Awesome work @woehrl01! thanks.

**Pull request merged**